alg/crc32c_sse42.c

```diff
@@ -0,0 +1,54 @@
+#include "cpusupport.h"
+#ifdef CPUSUPPORT_X86_CRC32
+
+#include <smmintrin.h>
+
+#include "crc32c_sse42.h"
+
+/**
+ * CRC32C_Update_SSE42(state, buf, len):
+ * Feed ${len} bytes from the buffer ${buf} into the CRC32C whose state is
+ * ${state}.  This implentation uses x86 SSE4.2 instructions, and should only
+ * be used if CPUSUPPORT_X86_CRC32 is defined and cpusupport_x86_crc32 returns
+ * nonzero.  ${len} must be greater than, or equal to, 8.
+ */
+uint32_t
+CRC32C_Update_SSE42(uint32_t state, const uint8_t * buf, size_t len)
+{
+	size_t pre_block;
+	size_t remaining_bytes;
+	size_t in_block;
+	size_t i = 0;
+
+	/*
+	 * Calculate how many bytes are before an aligned block.  Assume that
+	 * the (uintptr_t) value indicates the memory layout.
+	 */
+	pre_block = (8 - (uintptr_t)buf) & 7;
+
+	/* Calculate how many bytes are in aligned blocks. */
+	remaining_bytes = len - pre_block;
+	in_block = remaining_bytes - (remaining_bytes % 8);
+
+	/* Process bytes before the alignment. */
+	for (; i < pre_block; i++)
+		state = _mm_crc32_u8(state, buf[i]);
+
+	/*
+	 * Process blocks of 8.  It's ok if (i % 8 != 0), because we're
+	 * checking that (i < in_block), not (i != in_block).  For example,
+	 * if we start with i = 4 and want 1 block, we'll end up at i == 12.
+	 */
+	for (; i < in_block; i += 8) {
+		state = (uint32_t)_mm_crc32_u64(state,
+		    *(const unsigned long *)&buf[i]);
```

Comment:
Can we change this to `*(const uint64_t *)&buf[i]` ?  It seems that different headers spell this differently (clang has `unsigned long long` on my FreeBSD/amd64 system, the intel intrinsics guide says `unsigned __int64`, etc.) but it seems that `uint64_t` *should* work everywhere.

As it is right now, I worry that there will be some weird OS/compiler where `long` isn't 64 bits...

Reply:
Look ok on my system.  I agree that `long != 64` bits is a concern, which is why I find the Intel instrinsics often mindboggling.  (for example, `_mm_crc32_u32` takes `unsigned int` instead of `__uint32`)